File: capacitor-plugins/capacitor-native/ios/Sources/AonsokuNativePlugin/Debug/NativeLogger.swift
import Foundation

final class NativeLogger {
    static let shared = NativeLogger()

    struct Entry {
        let timestamp: Date
        let level: Level
        let message: String
        let source: String

        enum Level: String, CaseIterable {
            case debug, info, warn, error
        }
    }

    private var buckets: [String: [Entry]] = [:]
    private let maxEntriesPerSource = 200
    private let queue = DispatchQueue(label: "com.aonsoku.NativeLogger")

    private init() {}

    func log(_ level: Entry.Level, _ message: String, source: String = "") {
        queue.async {
            let entry = Entry(timestamp: Date(), level: level, message: message, source: source)
            let key = source.isEmpty ? "_default" : source
            var bucket = self.buckets[key] ?? []
            if bucket.count >= self.maxEntriesPerSource {
                bucket.removeFirst()
            }
            bucket.append(entry)
            self.buckets[key] = bucket
        }
    }

    func debug(_ message: String, source: String = "") {
        log(.debug, message, source: source)
    }

    func info(_ message: String, source: String = "") {
        log(.info, message, source: source)
    }

    func warn(_ message: String, source: String = "") {
        log(.warn, message, source: source)
    }

    func error(_ message: String, source: String = "") {
        log(.error, message, source: source)
    }

    func getEntries() -> [Entry] {
        queue.sync {
            buckets.values.flatMap { $0 }.sorted { $0.timestamp < $1.timestamp }
        }
    }

    func clear() {
        queue.async { self.buckets.removeAll() }
    }
}

File: capacitor-plugins/capacitor-native/android/src/main/java/github/realtvop/aonsoku/plugins/debug/NativeLogger.kt
package github.realtvop.aonsoku.plugins.debug

import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

object NativeLogger {
    data class Entry(
        val timestamp: Long,
        val level: Level,
        val message: String,
        val source: String
    ) {
        enum class Level { DEBUG, INFO, WARN, ERROR }
    }

    private const val MAX_ENTRIES_PER_SOURCE = 200
    private val buckets = mutableMapOf<String, MutableList<Entry>>()
    private val lock = ReentrantReadWriteLock()

    fun log(level: Entry.Level, message: String, source: String = "") {
        val entry = Entry(System.currentTimeMillis(), level, message, source)
        val key = source.ifEmpty { "_default" }
        lock.write {
            val bucket = buckets.getOrPut(key) { mutableListOf() }
            if (bucket.size >= MAX_ENTRIES_PER_SOURCE) {
                bucket.removeFirst()
            }
            bucket.add(entry)
        }
        val tag = "Aonsoku/${source.ifEmpty { "Default" }}"
        when (level) {
            Entry.Level.DEBUG -> android.util.Log.d(tag, message)
            Entry.Level.INFO -> android.util.Log.i(tag, message)
            Entry.Level.WARN -> android.util.Log.w(tag, message)
            Entry.Level.ERROR -> android.util.Log.e(tag, message)
        }
    }

    fun debug(message: String, source: String = "") = log(Entry.Level.DEBUG, message, source)
    fun info(message: String, source: String = "") = log(Entry.Level.INFO, message, source)
    fun warn(message: String, source: String = "") = log(Entry.Level.WARN, message, source)
    fun error(message: String, source: String = "") = log(Entry.Level.ERROR, message, source)

    fun getEntries(): List<Entry> {
        return lock.read {
            buckets.values.flatten().sortedBy { it.timestamp }
        }
    }

    fun clear() {
        lock.write { buckets.clear() }
    }
}
